### PR TITLE
#19219 correct recognize Trino varchar data type

### LINKED ISSUE #19219: Unable to import csv/txt file containing varchar data type into trino database
[###](url) Description

When importing a txt or csv file containing varchar data types, DBeaver will error out during import with "Object parameter [column value] binding not supported.

image

txt file used as example
image

Data Transfer Import Settings
image

Metadata structure
image

Data Load Settings
image

Error
image
hubusercontent.com/24360724/221951869-582bff85-b7ab-4753-ac2e-14c51dd34e75.png)

### DBeaver Version

Community 22.3.5

### Operating System

Windows 10 Enterprise

### Database and driver

Trino 383

### Steps to reproduce

1. Right click schema and click import data
2. Select CSV as source and click next
3. Select a txt file and click Next
4. Click next in the table mapping
5. Click Next  in Data load settings
6. Click next in Confirm
7. Click Proceed
8. Error will pop up regarding Object parameter binding not supported

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by LonwoLonwo**
Changes are added only for the Varchar data type in Trino. You can just observe these changes in the Data Types folder.

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/jdbc/cache/JDBCBasicDataTypeCache.java [hunk 1]
```diff
@@ -67,8 +68,10 @@ protected OBJECT fetchObject(@NotNull JDBCSession session, @NotNull OWNER owner,
             return null;
         }
         int valueType = JDBCUtils.safeGetInt(dbResult, JDBCConstants.DATA_TYPE);
-        // Check for bad value type for strings: #494
-        if (valueType == Types.BINARY && (name.contains("varchar") || name.contains("VARCHAR"))) {
+        // Check for bad value type for strings: #494 && 19219
+        if ((valueType == Types.BINARY && (name.contains(SQLConstants.DATA_TYPE_VARCHAR) || name.contains("VARCHAR"))) ||
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/jdbc/cache/JDBCBasicDataTypeCache.java]: Very strange check. Just do it in a case insensitive way

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/jdbc/cache/JDBCBasicDataTypeCache.java [hunk 2]
```diff
@@ -67,8 +68,10 @@ protected OBJECT fetchObject(@NotNull JDBCSession session, @NotNull OWNER owner,
             return null;
         }
         int valueType = JDBCUtils.safeGetInt(dbResult, JDBCConstants.DATA_TYPE);
-        // Check for bad value type for strings: #494
-        if (valueType == Types.BINARY && (name.contains("varchar") || name.contains("VARCHAR"))) {
+        // Check for bad value type for strings: #494 && 19219
+        if ((valueType == Types.BINARY  && name.toUpperCase(Locale.ENGLISH).contains("VARCHAR")) ||
+            valueType == Types.JAVA_OBJECT && SQLConstants.DATA_TYPE_VARCHAR.equalsIgnoreCase(name)
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/jdbc/cache/JDBCBasicDataTypeCache.java]: Are separate checks required for different types of data? Do we need to check if the name **contains** word VARCHAR only if its type is `Types.BINARY`?

```suggestion
        if ((valueType == Types.BINARY || valueType == Types.JAVA_OBJECT) && name.toUpperCase(Locale.ENGLISH).contains("VARCHAR"))

```

Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/jdbc/cache/JDBCBasicDataTypeCache.java]: yes. for historical reason.

Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/impl/jdbc/cache/JDBCBasicDataTypeCache.java]: Then let's be good boys and clarify our intentions:

```suggestion
            (valueType == Types.JAVA_OBJECT && SQLConstants.DATA_TYPE_VARCHAR.equalsIgnoreCase(name))

```

**Review by ShadelessFox — APPROVED** [review]

**Comment by MashaKorax** [comment]
verified

**Review by serge-rider — APPROVED** [review]

**Pull request merged**